core/src/main/java/io/grpc/ServerImpl.java

```diff
@@ -502,20 +508,27 @@ public ServerStreamListenerImpl(ServerCall.Listener<ReqT> listener, Future<?> ti
       }
 
       @Override
-      public void messageRead(final InputStream message) {
-        try {
-          if (cancelled) {
-            return;
-          }
-
-          listener.onPayload(method.parseRequest(message));
-        } finally {
+      public void messagesRead(final Iterator<InputStream> messages) {
+        RuntimeException exception = null;
+        while (messages.hasNext()) {
+          InputStream message = messages.next();
           try {
-            message.close();
-          } catch (IOException e) {
-            throw new RuntimeException(e);
+            if (!cancelled) {
+              listener.onPayload(method.parseRequest(message));
+            }
+          } finally {
+            try {
+              message.close();
+            } catch (IOException e) {
```

Comment:
One way to do that would be to invert the relationship a little and have something like

void messagesRead(MessageSource source) {
   source.drainTo(messageSink);
}

One of the reasons we might want a pattern like that is it might allow for some cleaner synchronization with the network thread and eliminate some of the over-notification that you see in the code currently.